docs/docs/resource-handling-and-service-workers.md

```diff
@@ -0,0 +1,71 @@
+
+Another current problem is that we may start fetching the resources for a page before the service worker has finished installing, but finish fetching them all after it has installed - this could cause a page's path to be whitelisted even if some of its resources haven't been cached (since Gatsby assumes the service worker was installed at the start of fetching resources, if it was installed at the end).
+
+## Gatsby Core
+
+### Resource loader (`loader.js`)
+
+There are two functions which perform a similar but distinct role in this file: `enqueue` and `getResourcesForPathname`. The former of these, `enqueue`, is designed to speed up navigation by prefetching resources for a page, before we need to display the page, and hence it doesn't return anything. On the other hand, `getResourcesForPathname` is used when we need the resources right now, usually in order to display the page, and therefore it fetches with higher priority than `enqueue` as well as returning them. Another difference between the two is that `getResourcesForPathname` returns the resources for the 404 page if the specified page doesn't exist.
+
+In the future, we could refactor these into a single function which takes parameters for whether or not to return the 404 page if the specified page is missing, and for whether to fetch with high or low priority.
+
+### EnsureResources (`ensure-resources.js`)
+
+The `EnsureResources` component exists to (as its name suggest) ensure that we have the resources for a specified page, and load them if we currently don't. There are some valid reasons why we might not have resources at the time of navigation:
+
+1. 404s resulting from site-internal links, when the user has not created a custom 404 page with `404.js`
+2. When visiting a page (which hasn't been visited before), the site may have been updated since our last navigation and so the resources to the unvisited page may have changed location
+3. Ad blockers may block access to certain paths in which case we're unable to load the JS resources for a page
+
+Here is how the `EnsureResources` component handles each of these scenarios:
+
+1. On non-initial renders, reload upon missing resources so that the browser can load the server's 404 page
+
+2. On non-initial renders, reload upon missing resources so that the browser can load the latest page
+
+   On initial renders, flag the page as failed and throw an error to prevent rendering a blank page (static HTML will suffice), then reload it once the service worker updates
+
+3. On initial renders, throw an error to prevent rendering a blank page (static HTML will suffice)
```

Comment:
Should this line be removed in favour of the previous line?

Reply:
This applies to 2 and 3 which is why it's duplicated - I thought it would be clearest to write it twice so that I can keep the numbered list format, although maybe it'll be clearer that it's intentional if I swap the order of the points under number 2.